lib/internal/modules/esm/importmap.js

```diff
@@ -0,0 +1,135 @@
+'use strict';
+const path = require('path');
+const { readFileSync } = require('fs');
+const { pathToFileURL } = require('internal/url');
+const { setOwnProperty } = require('internal/util');
+const { JSONParse, ObjectEntries, SafeMap } = primordials;
+// @TODO add error codes
+// const { } = require('internal/errors').codes;
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  static load(filepath) {
+    let str;
+    try {
+      str = readFileSync(filepath);
+    } catch (err) {
+      if (err.code !== 'ENOENT') {
+        setOwnProperty(err, 'importMapPath', `${filepath}`);
+        throw err;
+      }
+      return;
+    }
+
+    let json;
+    try {
+      json = JSONParse(str);
+    } catch (err) {
+      setOwnProperty(err, 'importMap', `${str}`);
+      throw err;
+    }
+
+    return new ImportMap(json, pathToFileURL(path.dirname(filepath)));
```

Comment:
```suggestion
    return new ImportMap(json, pathToFileURL(filepath));
```

Reply:
That is the base directory to resolve against. We need to remove `/importmap.json` from the end somewhere, I can do it in the constructor if you like, but it needs to happen somewhere.

Reply:
My point is that it probably doesn't need to be removed.

Reply:
Removing it did break the path resolution. For example this means that when we construct relative URL's to the import map we get paths like `test/fixtures/es-module-loaders/importmaps/simple.json/node_modules/foo/index.mjs`. That said, maybe I can find a way around that, I guess I just want to know more about why you suggested I remove this. Is there something other than the maybe you were thinking of here for why it should not be necessary?

Reply:
`new URL('someFile', 'file:///foo/importmap.json')` and `new URL('someFile', 'file:///foo/')` both resolve to the same URL (`file:///foo/someFile`), so I'm surprised calling dirname is not pure overhead 🤔 maybe there's a bug somewhere, hopefully not in the Node.js `URL` implementation.

Reply:
Maybe something to do with the implementation in `pathToFileURL`?

Reply:
The fact that this PR needs `pathToFileURL` at all is surprising, I thought that the import map proposal, coming from the web, would deal only with URLs, so it’s weird that we end up with paths that we need to convert to URLs. I admit that I haven’t read the spec text carefully, so maybe it’s correct, but I think it’s fair to say it’s quite surprising to find paths in a URL world like ESM. It might be a good idea to add comments explaining why we have paths and where there are coming from.

Reply:
I mean maybe my expectations based on examples in the ecosystem, the spec, and my personal bias against anything other than file paths used in this context is getting in the way. Where is there a hard and fast definition of what "url like" means? Because what I tried initially was not parsed by `new URL`, it errored out.

Reply:
Yeah, it might be that a few other things I was doing wrong earlier made wrong assumptions. I might be able to work out just removing those calls. Let me mess with that a bit.

Reply:
Oh yeah! I remember now one of the first things I hit, and likely might be to your original point of not needing to use dirname.

Reply:
Glad you pushed on this point. I think I got it sorted out and it was just a few of the examples I pulled from initially mixed usage and instead of really trying to convert everything to URL instances I looked for other ways. Will update the PR in a little bit.